internal/wasm/module_instance.go

```diff
@@ -146,6 +146,12 @@ func (m *ModuleInstance) setExitCode(exitCode uint32, flag exitCodeFlag) bool {
 // ensureResourcesClosed ensures that resources assigned to ModuleInstance is released.
 // Multiple calls to this function is safe.
 func (m *ModuleInstance) ensureResourcesClosed(ctx context.Context) (err error) {
+	if notification := m.CloseNotification; notification != nil { // experimental
+		closed := atomic.LoadUint64(&m.Closed)
+		notification.OnClose(ctx, uint32(closed>>32))
+		m.CloseNotification = nil
+	}
```

Comment:
shouldn't this be guarded by sync.Once?

Reply:
yep! ensureResourcesClosed is externally guarded atomically to make sure it is only called once

Reply:
I'll add the missing comment

Reply:
aha!